### PR TITLE
Empty ISD handling

### LINKED ISSUE #6: HRM and screen clearing after empty ISD
I've tried to run one example of subtitles in https://github.com/sandflow/imscHRM and ran into the following problem.  Consider the following case:
```
<p xml:id="subtitle3" begin="20.062s" end="22.940s" region="region1" style="style1">このトランプ<br/>おしっこの臭いがする</p>
<p xml:id="subtitle4" begin="23.023s" end="24.024" region="region1" style="style1">嗅いでみて</p>
```
imscHRM produces:
```
ERROR:imschrm.hrm:Rendering time exceeded at 23.023s (doc w3c/imsc#7)
  available time: 0.083s | HRM time: 0.101
  Glyph copy count: 0 | render count: 5 | Background draw count: 0

Validation failed
```
The XML produces at least 3 ISD:
* one from 20.062s to 22.940s
* one from 22.940s to 23.023s (i.e. 2-frame duration at 24fps), empty ISD
* one from 23.023s to 23.023s

After discussion with @palemieux it seems that the problem is that the HRM always assumes that a clear operation has to be performed prior to painting the new ISD text. The IMSC spec says:

> DUR(En) = S(En) / BDraw + DURT(En) + DURI(En)
> S(En) = CLEAR(En) + PAINT(En )
> where CLEAR(E0) = 0 and CLEAR(En | n > 0) = 1, i.e. the Root Container Region in its entirety.

In the above case, ISD2 already has cleared the screen, therefore there should be no need to clear the screen again to paint ISD3.

### PR DESCRIPTION
**Pull request by palemieux**
Closes #6

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/imsc-hrm/pull/9.html" title="Last updated on Sep 30, 2021, 3:50 AM UTC (c35df40)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/imsc-hrm/9/129d887...c35df40.html" title="Last updated on Sep 30, 2021, 3:50 AM UTC (c35df40)">Diff</a>

### CODE REVIEW

File: spec/imsc-hrm.html [hunk 1]
```diff
@@ -294,11 +292,16 @@ <h2>General</h2>
       <p>The Presentation Compositor SHALL start rendering E<sub>n</sub>:</p>
 
       <ul>
-        <li>at the presentation time of E<sub>0</sub> minus Initial Painting Delay (IPD), if n = 0; or</li>
-
-        <li>at the presentation time of E<sub>n-1</sub>, if n &gt; 0.</li>
+        <li>at the presentation time of E<sub>m</sub>, where E<sub>m</sub> is the closest ISD for which m&lt;n and E<sub>m-1</sub>
```
Comment (cconcolato) [inline spec/imsc-hrm.html:295]: Not sure the previous text was wrong. Is there a problem to render an empty ISD?

Comment (palemieux) [reply spec/imsc-hrm.html:295]: In the previous text, the time available to render ISD_n was the duration of ISD_(n-1). This is a problem if ISD_(n-1)...(n-m) are all empty and the duration of ISD_(n-1), which is empty, is very short.

**Review by cconcolato — APPROVED** [review]

File: spec/imsc-hrm.html [hunk 2]
```diff
@@ -280,9 +280,7 @@ <h2>General</h2>
       document</a> E<sub>n</sub> using the following steps in order:</p>
 
       <ol>
-        <li>clear the pixels, except for the first <a>intermediate synchronic document</a> E<sub>0</sub> for the which the pixels
-        of P<sub>0</sub> SHALL be assumed to have been cleared;
-        </li>
+        <li>clear the pixels of the entire <a>Root Container Region</a>, unless n=0 or E<sub>n-1</sub> was empty;</li>
```
Comment (nigelmegitt) [inline spec/imsc-hrm.html:283]: Do we need to make any provision for consecutive document playback, e.g. what happens in a segmented delivery scenario? If we do, possibly `n=0` would need to be adjusted, or we would need some ISD comparison and deduplication. Possibly a separate issue.

Comment (nigelmegitt) [reply spec/imsc-hrm.html:283]: I don't think it's entirely clear what "empty" means here.

File: spec/imsc-hrm.html [hunk 3]
```diff
@@ -294,11 +292,16 @@ <h2>General</h2>
       <p>The Presentation Compositor SHALL start rendering E<sub>n</sub>:</p>
 
       <ul>
-        <li>at the presentation time of E<sub>0</sub> minus Initial Painting Delay (IPD), if n = 0; or</li>
-
-        <li>at the presentation time of E<sub>n-1</sub>, if n &gt; 0.</li>
+        <li>at the presentation time of E<sub>m</sub>, where E<sub>m</sub> is the closest ISD for which m&lt;n and E<sub>m-1</sub>
+        is not empty, if E<sub>m</sub> exists and the presentation time of E<sub>n</sub> minus that of E<sub>m</sub> is less than
```
Comment (nigelmegitt) [inline spec/imsc-hrm.html:296]: This is really hard text to parse. In particular, I'm not sure what the `if` applies to.

One possible reading is that we only attempt to render E<sub>n</sub> if it's presentation time is >= 1 second (== IPD) later than that of E<sub>m</sub>? That doesn't seem desirable. It's frequently the case that adjacent ISDs are activated <1s apart.

Another is that if the two ISDs are >=1s apart, then we begin rendering at E<sub>0</sub> rather than at the presentation time of E<sub>n</sub> - IPD, which is what I'd expect.

In other words, _every_ ISD should be renderable in <= IPD, right?

File: spec/imsc-hrm.html (hunk 1 above)
Comment (nigelmegitt) [inline spec/imsc-hrm.html:295]: Shouldn't it be E<sub>m+1</sub> is not empty? Otherwise if the sequence is:

E<sub>n-x-1</sub>: Empty
E<sub>n-x</sub>: Has visible content
E<sub>n-x+1</sub>: Empty
E<sub>...</sub>: [maybe more empty ones]
E<sub>n</sub>: Has visible content

We want m == n-x here, I think. But E<sub>n-x-1</sub> is empty, so the current text excludes m == n-x which is an unexpected result.

**Review by nigelmegitt — CHANGES_REQUESTED** [review]
I'm not actually sure we're all agreed on what the intended result is here; in any case I think we need to define "empty" and make it easier to understand the rules for when we consider rendering to begin for any ISD.

File: spec/imsc-hrm.html (hunk 2 above)
Comment (palemieux) [reply spec/imsc-hrm.html:283]: > I don't think it's entirely clear what "empty" means here.

Will add a definition.

Comment (palemieux) [reply spec/imsc-hrm.html:283]: >  or we would need some ISD comparison and deduplication. Possibly a separate issue.

Please file an issue

File: spec/imsc-hrm.html (hunk 3 above)
Comment (palemieux) [reply spec/imsc-hrm.html:296]: > In other words, every ISD should be renderable in <= IPD, right?

Yes.

File: spec/imsc-hrm.html (hunk 1 above)
Comment (palemieux) [reply spec/imsc-hrm.html:295]: Urgh. Thanks for catching this. Will fix.

Comment (palemieux) [reply spec/imsc-hrm.html:295]: Actually we want `m == n-x+1`: assuming there is no sequences of empty ISDs, the rendering of `E_n` should start at the presentation time of `E_(n-1)`, even if `E_(n-1)` is empty.

File: spec/imsc-hrm.html (hunk 2 above)
Comment (palemieux) [reply spec/imsc-hrm.html]: @nigelmegitt Definition for _empty ISD_ added at https://github.com/w3c/imsc-hrm/pull/9/commits/b17fe84f4bb21cc5a8da0bb2cef1a499f509abaa

File: spec/imsc-hrm.html (hunk 1 above)
Comment (nigelmegitt) [reply spec/imsc-hrm.html:297]: I thought that's what the current text says, that causes the problem when there's a 2 frame (say) empty ISD, and that's what we wanted to resolve. That's why I thought the conclusion was that we want the "render timer" to begin at `n-x` rather than `n-x+1`.

File: spec/imsc-hrm.html [hunk 4]
```diff
@@ -229,6 +229,8 @@ <h2>Terms and Definitions</h2>
       <p><dfn>presented region</dfn>. See [[!imsc]].</p>
 
       <p><dfn>presented image</dfn>. See [[!imsc]].</p>
+      
+      <p><dfn>empty ISD</dfn>. an <a>Intermediate Synchronic Document</a> with at least one <a>presented region</a>.</p>
```
[the remark this answers is not in the document]
Comment (palemieux) [reply spec/imsc-hrm.html:233]: Doh. It should say _with no presented region_.

Fixed.

File: spec/imsc-hrm.html (hunk 1 above)
Comment (palemieux) [reply spec/imsc-hrm.html:297]: Given the following sequence:

```
E5: Empty
E6: Has visible content
E7: Empty
E8: Empty
E9: Empty
E10: Has visible content
```

We want the rendering of E10 to start at the presentation time of E7 and not E9.

Comment (nigelmegitt) [reply spec/imsc-hrm.html:297]: This is where we disagree, perhaps: I think we need it to be E6 not E7.

Comment (palemieux) [reply spec/imsc-hrm.html:297]: As currently specified, the rendering of document E_n starts at the presentation time of document E_(n-1), even if E_(n-1) is empty. This is necessary since there is a cost to E_(n-1), even if E_(n-1) is empty: the root container has to be cleared. Do you agree?

If so, all that we are doing is making sure that multiple empty ISDs in sequence are considered as one.

Comment (nigelmegitt) [reply spec/imsc-hrm.html:297]: Yes, right, but I don't think that solves the problem that's been presented, which is that a single 2 frame empty ISD causes the following ISD to fail the HRM.

I think we need to imagine that "clearing all the pixels" is an activity that can be done independently of "preparing presentation of a non-empty ISD", and that it has a known fixed cost.

Comment (palemieux) [reply spec/imsc-hrm.html:297]: >Yes, right, but I don't think that solves the problem that's been presented, 

The problem was two-fold based on the files I have seen:

* E10 was forced to clear the root container even though it had been cleared by E7
* E10 only had the duration of E9 to render, even though rendering could have started at E7 -- this is particularly acute if E9 is very short.

This PR is intended to address the two issues above. Makes sense?

Comment (nigelmegitt) [reply spec/imsc-hrm.html:297]: Yes, makes sense that it would resolve those issues, but what doesn't make sense is how we get to a situation where there are multiple consecutive empty ISDs, given that the `[resolve timing]` procedure is only supposed to generate time coordinates where some element becomes active or inactive. If two consecutive ISDs are both empty, doesn't that imply that no element became active between the first and second? And since the ISDs are empty, there wasn't anything to become inactive that was already active.

Is it not correct that we need a solution to short (single) empty ISDs causing files to fail HRM validation?

Comment (palemieux) [reply spec/imsc-hrm.html:297]: > , but what doesn't make sense is how we get to a situation where there are multiple consecutive empty ISDs,

Two scenarios come to mind:

* the `[resolve timing]` does not preclude generating multiple empty ISDs in sequence. For example, `set` elements can generate time coordinates without causing visual differences and without being present in the ISD (by definition):

```
<region xml:id="r1" begin="0s" end="1s" tts:background="transparent">
   <set tts:background="transparent" begin="0.5s" end="1s">
</region>
```

* there is a temporal gap between documents in a sequence of documents, resulting effectively in an empty ISD.

There might be more.

Comment (palemieux) [reply spec/imsc-hrm.html]: @nigelmegitt  added illustration

**Review by nigelmegitt — APPROVED** [review]
Thanks for the updates and fixes @palemieux , looks good to me.
